### PR TITLE
Extend usage for olora finetune script

### PR DESCRIPTION
**Pull request by jiqing-feng**
I compared a lot of lora finetune script and found this script is the best, it's so clear and easy to understand. So I want to extend the usage for this script to support more platforms and also DDP with minimum changes.

I currently test the script on cuda and cpu with 4 DDP on opt and llama-3. It shows reasonable results. 

It would be great if this PR can be merged so we can apply it on more models and devices.

### CODE REVIEW

File: examples/olora_finetuning/README.md [hunk 1]
```diff
@@ -39,6 +39,15 @@ OLoRA also supports quantization. To use 4-bit quantization try:
 ```bash
 python3 examples/olora_finetuning/olora_finetuning.py --base_model facebook/opt-350m --quantize
 ```
+or you can just pass a quantized model without the quantize flag.
+
+If you want to run DDP by accelerate, please run `accelerate config` to set your ddp config, and run:
```
Comment (BenjaminBossan) [inline examples/olora_finetuning/README.md:44]: Let's add a link to the accelerate docs -- not everyone may know this package.

File: examples/olora_finetuning/olora_finetuning.py [hunk 2]
```diff
@@ -43,23 +44,32 @@ def train(
     lora_alpha: int = 16,
     lora_dropout: float = 0.05,
     lora_target_modules: List[str] = None,
+    seed: int = None,
```
Comment (BenjaminBossan) [inline examples/olora_finetuning/olora_finetuning.py:47]:
```suggestion
    seed: Optional[int] = None,
```
(`Optional` needs to be imported)

File: examples/olora_finetuning/README.md [hunk 3]
```diff
@@ -39,6 +39,15 @@ OLoRA also supports quantization. To use 4-bit quantization try:
...
+```bash
+accelerate launch examples/olora_finetuning/olora_finetuning.py --base_model facebook/opt-350m
+```
+please add `--device_map cpu` if you want to run finetune on CPU.
+
+If you want to train a quantized model like AWQ and GPTQ which do not support olora init method, please pass `--init_lora_weights gaussian`.
```
Comment (BenjaminBossan) [inline examples/olora_finetuning/README.md:50]: Would it be possible to use AWQ or GPTQ right now? If the user passes `--quantize`, the usage of bitsandbytes is hard-coded.

File: examples/olora_finetuning/olora_finetuning.py [hunk 4]
```diff
@@ -43,23 +44,32 @@ def train(
...
-        base_model,
-        device_map=device_map,
-        quantization_config=BitsAndBytesConfig(
+    # Set device_map to the right place when enabling DDP.
+    world_size = int(os.environ.get("WORLD_SIZE", 0)) or int(os.environ.get("PMI_SIZE", 0))
+    if world_size > 1 and device_map != "cpu":
+        from accelerate import Accelerator
+        device_map = {'': Accelerator().process_index}
+    # Set seed
+    if seed:
```
Comment (BenjaminBossan) [inline examples/olora_finetuning/olora_finetuning.py:57]:
```suggestion
    if seed is not None:
```

Otherwise, passing `0` would not set the seed.

File: examples/olora_finetuning/olora_finetuning.py [hunk 5]
```diff
@@ -43,23 +44,32 @@ def train(
...
-        if quantize
-        else None,
-        torch_dtype=torch.float16,
-    )
+    model = AutoModelForCausalLM.from_pretrained(base_model, **model_kwargs)
 
     tokenizer = AutoTokenizer.from_pretrained(base_model, trust_remote_code=True)
+    # For some tokenizer with no pad token like llama
+    if tokenizer.pad_token is None:
+        tokenizer.pad_token_id = 0
```
Comment (BenjaminBossan) [inline examples/olora_finetuning/olora_finetuning.py:72]: Can we just hard-code 0 here? How about: `tokenizer.pad_token = tokenizer.eos_token`

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for making these additions to the OLoRA fine-tuning script. Overall I think these are good additions, I just have a few small comments.

Also pinging @tokenizer-decode in case you also want to review but it's not necessary for this change, as OLoRA is not directly involved.

**Comment by tokenizer-decode** [comment]
Seems like good addition to me. I haven't run the script, but it looks good at first glance. You can ping me if anything is needed. Thanks btw @jiqing-feng

File: examples/olora_finetuning/README.md (hunk 3 above)
Comment (jiqing-feng) [reply examples/olora_finetuning/README.md]: Yes, we can just pass a quantized model. I have updated it in the readme.

**Comment by jiqing-feng** [comment]
Hi @BenjaminBossan . I have fixed all your comments, please review the new changes. Thanks!

File: examples/olora_finetuning/olora_finetuning.py [hunk 6]
```diff
@@ -122,6 +131,7 @@ def generate_and_tokenize_prompt(example):
             output_dir=output_dir,
             save_total_limit=3,
             load_best_model_at_end=True,
+            ddp_find_unused_parameters=False if world_size > 1 else True,
```
[the remark this answers is not in the document]
Comment (jiqing-feng) [reply examples/olora_finetuning/olora_finetuning.py:134]: Without this parameter, it will print the following warning when runnnig DDP:
```
[rank1]:[W108 19:06:54.522318540 reducer.cpp:1400] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters in the forward pass. This flag results in an extra traversal of the autograd graph every iteration,  which can adversely affect performance. If your model indeed never has any unused parameters in the forward pass, consider turning this flag off. Note that this warning may be a false positive if your model has flow control causing later iterations to have unused parameters. (function operator())
[rank2]:[W108 19:06:54.522318531 reducer.cpp:1400] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters in the forward pass. This flag results in an extra traversal of the autograd graph every iteration,  which can adversely affect performance. If your model indeed never has any unused parameters in the forward pass, consider turning this flag off. Note that this warning may be a false positive if your model has flow control causing later iterations to have unused parameters. (function operator())
[rank0]:[W108 19:06:54.523105147 reducer.cpp:1400] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters in the forward pass. This flag results in an extra traversal of the autograd graph every iteration,  which can adversely affect performance. If your model indeed never has any unused parameters in the forward pass, consider turning this flag off. Note that this warning may be a false positive if your model has flow control causing later iterations to have unused parameters. (function operator())
[rank3]:[W108 19:06:54.523123410 reducer.cpp:1400] Warning: find_unused_parameters=True was specified in DDP constructor, but did not find any unused parameters in the forward pass. This flag results in an extra traversal of the autograd graph every iteration,  which can adversely affect performance. If your model indeed never has any unused parameters in the forward pass, consider turning this flag off. Note that this warning may be a false positive if your model has flow control causing later iterations to have unused parameters. (function operator())

```

[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply examples/olora_finetuning/olora_finetuning.py:134]: I mean changing the line to: `ddp_find_unused_parameters=False if world_size > 1 else None`. Is that what you tested?

[the remark this answers is not in the document]
Comment (jiqing-feng) [reply examples/olora_finetuning/olora_finetuning.py:134]: The parameter should be setted to True [here](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L2051-L2058)

[the remark this answers is not in the document]
Comment (jiqing-feng) [reply examples/olora_finetuning/olora_finetuning.py:134]: Right, it should be None if no DDP.

[the remark this answers is not in the document]
Comment (jiqing-feng) [reply examples/olora_finetuning/olora_finetuning.py]: Fixed

**Comment by jiqing-feng** [comment]
Please trigger the test again. Thanks!

**Comment by jiqing-feng** [comment]
Format fixed.

**Comment by jiqing-feng** [comment]
The failed tests do not seem to be related to my changes.

**Review by BenjaminBossan — APPROVED** [review]
The changes LGTM, thanks.

Failing CI is indeed unrelated, I'll merge.
